src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java

```diff
@@ -0,0 +1,374 @@
+
+/**
+ * A simple predicate on a columns expression (e.g. columnA = X).
+ */
+public final class SimpleRestriction implements SingleRestriction
+{
+    /**
+     * The columns expression to which the restriction applies.
+     */
+    private final ColumnsExpression columnsExpression;
+
+    /**
+     * The operator
+     */
+    private final Operator operator;
+
+    /**
+     * The values
+     */
+    private final Terms values;
+
+    public SimpleRestriction(ColumnsExpression columnsExpression, Operator operator, Terms values)
+    {
+        this.columnsExpression = columnsExpression;
+        this.operator = operator;
+        this.values = values;
+    }
+
+    @Override
+    public boolean isOnToken()
+    {
+        return columnsExpression.kind() == ColumnsExpression.Kind.TOKEN;
+    }
+
+    @Override
+    public ColumnMetadata firstColumn()
+    {
+        return columnsExpression.firstColumn();
+    }
+
+    @Override
+    public ColumnMetadata lastColumn()
+    {
+        return columnsExpression.lastColumn();
+    }
+
+    @Override
+    public List<ColumnMetadata> columns()
+    {
+        return columnsExpression.columns();
+    }
+
+    @Override
+    public boolean isMultiColumn()
+    {
+        return columnsExpression.kind() == ColumnsExpression.Kind.MULTI_COLUMN;
+    }
+
+    @Override
+    public boolean isColumnLevel()
+    {
+        return columnsExpression.isColumnLevelExpression();
+    }
+
+    public Operator operator()
+    {
+        return operator;
+    }
+
+    @Override
+    public boolean isANN()
+    {
+        return operator == Operator.ANN;
+    }
+
+    @Override
+    public boolean isEQ()
+    {
+        return operator == Operator.EQ;
+    }
+
+    @Override
+    public boolean isSlice()
+    {
+        return operator.isSlice();
+    }
+
+    @Override
+    public boolean isIN()
+    {
+        return operator == Operator.IN;
+    }
+
+    /**
+     * Checks if this restriction operator is a CONTAINS, CONTAINS_KEY or is an equality on a map element.
+     * @return {@code true} if the restriction operator is one of the contains operations, {@code false} otherwise.
+     */
+    public boolean isContains()
+    {
+        return operator == Operator.CONTAINS
+               || operator == Operator.CONTAINS_KEY
+               || columnsExpression.kind() == ColumnsExpression.Kind.MAP_ELEMENT;
+    }
+
+    @Override
+    public boolean needsFilteringOrIndexing()
+    {
+        // The need for filtering or indexing is a combination of columns expression type and operator
+        // Therefore, we have to take both into account.
+        return columnsExpression.kind() == ColumnsExpression.Kind.MAP_ELEMENT
+               || operator.requiresFilteringOrIndexingFor(columnsExpression.columnsKind());
+    }
+
+    @Override
+    public void addFunctionsTo(List<Function> functions)
+    {
+        columnsExpression.addFunctionsTo(functions);
+        values.addFunctionsTo(functions);
```

Comment:
The IS NOT NULL code was never called because the Restriction was never stored with the other restrictions: https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java#L205